src/main/scala/wasm/MiniWasmFX.scala

```diff
@@ -160,27 +161,31 @@ case class EvaluatorFX(module: ModuleInstance) {
         resume.k(List(), kont, List(), mkont, List())
       case Throw() =>
         val err :: newStack = stack
-        def kr(s: Stack, k1: Cont[Ans], t1: List[Cont[Ans]], m1: MCont[Ans], hs: Handlers[Ans]): Ans =
+        def kr(s: Stack, k1: Cont[Ans], t1: Trail[Ans], m1: MCont[Ans], hs: Handlers[Ans]): Ans =
           kont(s, List(), s1 => k1(s1, List(), m1))
         hs.head._2(List(err, ContV(kr)), initK[Ans], List(), mkont)
 
       // WasmFX effect handlers:
       case ContNew(ty) =>
         val RefFuncV(f) :: newStack = stack
-        def kr(s: Stack, k1: Cont[Ans], t1: List[Cont[Ans]], m1: MCont[Ans], hs: Handlers[Ans]): Ans = {
-          evalCall1(f, s, frame/*?*/, k1, List(), m1, List(), hs, false)
+        def kr(s: Stack, k1: Cont[Ans], t1: Trail[Ans], m1: MCont[Ans], hs: Handlers[Ans]): Ans = {
+          evalCall1(f, s, frame/*?*/, k1, t1, m1, List(), hs, false)
         }
-        kont(ContV(kr) :: newStack, List(), mkont)
+        kont(ContV(kr) :: newStack, trail, mkont)
       case Suspend(tagId) =>
         val FuncType(_, inps, out) = module.tags(tagId)
         val (inputs, restStack) = stack.splitAt(inps.size)
-        val kr = (s: Stack, k1: Cont[Ans], t1: List[Cont[Ans]], m1: MCont[Ans], hs: Handlers[Ans]) => {
+        val kr = (s: Stack, _: Cont[Ans], t1: Trail[Ans], m1: MCont[Ans], hs: Handlers[Ans]) => {
           // FIXME: handlers are lost here
-          kont(s ++ restStack, List(), s1 => k1(s1, List(), m1)) // mkont lost here
+          val index = trail.indexWhere { case (_, tags) => tags.contains(tagId) }
+          val newTrail = if (index >= 0) trail.take(index) else trail
+          kont(s ++ restStack, newTrail ++ t1, m1) // mkont lost here, and it's safe if we never modify it
         }
         val newStack = ContV(kr) :: inputs
         hs.find(_._1 == tagId) match {
-          case Some((_, handler)) => handler(newStack, initK[Ans], List(), mkont)
+          case Some((_, handler)) => 
```

Comment:
ideally `hs.find(_._1 == tagId)` and `tags.contains(tagId)` find the same `resume` clause